File: carouselview/src/main/java/com/elliecoding/carouselview/CarouselView.java
package com.elliecoding.carouselview;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSnapHelper;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SnapHelper;

import com.elliecoding.carouselview.enums.IndicatorAnimationType;
import com.elliecoding.carouselview.enums.OffsetType;
import com.rd.PageIndicatorView;
import com.rd.animation.type.AnimationType;

@SuppressWarnings("unused") // This is a library, external consumers will use the public methods
public class CarouselView extends FrameLayout {

    private final Context context;
    private PageIndicatorView pageIndicatorView;
    private RecyclerView carouselRecyclerView;
    private CarouselLinearLayoutManager layoutManager;
    private CarouselViewListener carouselViewListener;
    private CarouselScrollListener carouselScrollListener;
    private IndicatorAnimationType indicatorAnimationType;
    private CarouselViewAdapter adapter;
    private OffsetType offsetType;
    private SnapHelper snapHelper;
    private boolean enableSnapping;
    private boolean enableAutoPlay;
    private int autoPlayDelay;
    private Handler autoPlayHandler;
    private boolean scaleOnScroll;
    private int resource;
    private int size;
    private int spacing;
    private int currentItem;
    private boolean isResourceSet = false;

    public CarouselView(@NonNull Context context) {
        super(context);
        this.context = context;
        init(null);
    }

    public CarouselView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        init(attrs);
    }

    private void init(AttributeSet attributeSet) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View carouselView = inflater.inflate(R.layout.view_carousel, this);
        this.carouselRecyclerView = carouselView.findViewById(R.id.carouselRecyclerView);
        this.pageIndicatorView = carouselView.findViewById(R.id.pageIndicatorView);
        this.autoPlayHandler = new Handler(Looper.myLooper());

        carouselRecyclerView.setHasFixedSize(false);
        this.initializeAttributes(attributeSet);
    }

    private void initializeAttributes(AttributeSet attributeSet) {
        if (attributeSet != null) {
            TypedArray attributes = this.context.getTheme().obtainStyledAttributes(attributeSet,
                    R.styleable.CarouselView, 0, 0);
            this.enableSnapping(attributes.getBoolean(R.styleable.CarouselView_enableSnapping, true));
            this.setScaleOnScroll(attributes.getBoolean(R.styleable.CarouselView_scaleOnScroll, false));
            this.setAutoPlay(attributes.getBoolean(R.styleable.CarouselView_setAutoPlay, false));
            this.setAutoPlayDelay(attributes.getInteger(R.styleable.CarouselView_setAutoPlayDelay, 2500));
            this.setCarouselOffset(this.getOffset(attributes.getInteger(R.styleable.CarouselView_carouselOffset, 0)));
            int resourceId = attributes.getResourceId(R.styleable.CarouselView_resource, 0);
            if (resourceId != 0) {
                this.setResource(resourceId);
            }
            int indicatorSelectedColorResourceId =
                    attributes.getColor(R.styleable.CarouselView_indicatorSelectedColor, 0);
            int indicatorUnselectedColorResourceId =
                    attributes.getColor(R.styleable.CarouselView_indicatorUnselectedColor, 0);
            if (indicatorSelectedColorResourceId != 0) {
                this.setIndicatorSelectedColor(indicatorSelectedColorResourceId);
            }
            if (indicatorUnselectedColorResourceId != 0) {
                this.setIndicatorUnselectedColor(indicatorUnselectedColorResourceId);
            }
            this.setIndicatorAnimationType(this.getAnimation(attributes.getInteger(R.styleable.CarouselView_indicatorAnimationType, 0)));
            this.setIndicatorRadius(attributes.getInteger(R.styleable.CarouselView_indicatorRadius, 5));
            this.setIndicatorPadding(attributes.getInteger(R.styleable.CarouselView_indicatorPadding, 5));
            this.setSize(attributes.getInteger(R.styleable.CarouselView_size, 0));
            this.setSpacing(attributes.getInteger(R.styleable.CarouselView_spacing, 0));
            attributes.recycle();
        }
    }

    public void enableSnapping(boolean enable) {
        this.enableSnapping = enable;
    }

    public void hideIndicator(boolean hide) {
        if (hide) {
            this.pageIndicatorView.setVisibility(GONE);
        } else {
            this.pageIndicatorView.setVisibility(VISIBLE);
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        this.setAutoPlay(false);
    }

    private void setAdapter() {
        this.layoutManager = new CarouselLinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false);
        this.layoutManager.isOffsetStart(this.getCarouselOffset() == OffsetType.START);
        if (this.getScaleOnScroll()) this.layoutManager.setScaleOnScroll(true);
        carouselRecyclerView.setLayoutManager(this.layoutManager);
        this.adapter = new CarouselViewAdapter(getCarouselViewListener(), getResource(), getSize(),
                carouselRecyclerView, this.getSpacing(), this.getCarouselOffset() == OffsetType.CENTER);
        this.carouselRecyclerView.setAdapter(adapter);
        if (this.enableSnapping) {
            this.snapHelper.attachToRecyclerView(this.carouselRecyclerView);
        }
        this.setScrollListener();
        this.enableAutoPlay();
    }

    public CarouselViewAdapter getAdapter() {
        return adapter;
    }

    private void setScrollListener() {
        this.carouselRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                View centerView = snapHelper.findSnapView(layoutManager);
                int position = layoutManager.getPosition(centerView);
                if (carouselScrollListener != null) {
                    carouselScrollListener.onScrollStateChanged(recyclerView, newState, position);
                }
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    pageIndicatorView.setSelection(position);
                    setCurrentItem(position);
                }
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (carouselScrollListener != null) {
                    carouselScrollListener.onScrolled(recyclerView, dx, dy);
                }
            }
        });
    }

    public boolean getAutoPlay() {
        return this.enableAutoPlay;
    }

    public void setAutoPlay(boolean enableAutoPlay) {
        this.enableAutoPlay = enableAutoPlay;
    }

    public int getAutoPlayDelay() {
        return this.autoPlayDelay;
    }

    public void setAutoPlayDelay(int autoPlayDelay) {
        this.autoPlayDelay = autoPlayDelay;
    }

    private void enableAutoPlay() {
        autoPlayHandler.postDelayed(new Runnable() {
            public void run() {
                if (getAutoPlay()) {
                    if (getSize() - 1 == getCurrentItem()) {
                        setCurrentItem(0);
                    } else {
                        setCurrentItem(getCurrentItem() + 1);
                    }
                    autoPlayHandler.postDelayed(this, getAutoPlayDelay());
                }
            }
        }, getAutoPlayDelay());
    }

    public OffsetType getCarouselOffset() {
        return this.offsetType;
    }

    public void setCarouselOffset(OffsetType offsetType) {
        this.offsetType = offsetType;
        switch (offsetType) {
            case CENTER:
                this.snapHelper = new LinearSnapHelper();
                break;
            case START:
                this.snapHelper = new CarouselSnapHelper();
                break;
        }
    }

    public int getCurrentItem() {
        return this.currentItem;
    }

    public void setCurrentItem(int item) {
        if (item < 0) {
            this.currentItem = 0;
        } else if (item >= this.getSize()) {
            this.currentItem = this.getSize() - 1;
        } else {
            this.currentItem = item;
        }
        this.carouselRecyclerView.smoothScrollToPosition(this.currentItem);
    }

    public IndicatorAnimationType getIndicatorAnimationType() {
        return this.indicatorAnimationType;
    }

    public void setIndicatorAnimationType(IndicatorAnimationType indicatorAnimationType) {
        this.indicatorAnimationType = indicatorAnimationType;
        switch (indicatorAnimationType) {
            case DROP:
                this.pageIndicatorView.setAnimationType(AnimationType.DROP);
                break;
            case FILL:
                this.pageIndicatorView.setAnimationType(AnimationType.FILL);
                break;
            case NONE:
                this.pageIndicatorView.setAnimationType(AnimationType.NONE);
                break;
            case SWAP:
                this.pageIndicatorView.setAnimationType(AnimationType.SWAP);
                break;
            case WORM:
                this.pageIndicatorView.setAnimationType(AnimationType.WORM);
                break;
            case COLOR:
                this.pageIndicatorView.setAnimationType(AnimationType.COLOR);
                break;
            case SCALE:
                this.pageIndicatorView.setAnimationType(AnimationType.SCALE);
                break;
            case SLIDE:
                this.pageIndicatorView.setAnimationType(AnimationType.SLIDE);
                break;
            case THIN_WORM:
                this.pageIndicatorView.setAnimationType(AnimationType.THIN_WORM);
                break;
            case SCALE_DOWN:
                this.pageIndicatorView.setAnimationType(AnimationType.SCALE_DOWN);
        }
    }

    public int getIndicatorRadius() {
        return this.pageIndicatorView.getRadius();
    }

    public void setIndicatorRadius(int radius) {
        this.pageIndicatorView.setRadius(radius);
    }

    public int getIndicatorPadding() {
        return this.pageIndicatorView.getPadding();
    }

    public void setIndicatorPadding(int padding) {
        this.pageIndicatorView.setPadding(padding);
    }

    public int getIndicatorSelectedColor() {
        return this.pageIndicatorView.getSelectedColor();
    }

    public void setIndicatorSelectedColor(int color) {
        this.pageIndicatorView.setSelectedColor(color);
    }

    public int getIndicatorUnselectedColor() {
        return this.pageIndicatorView.getUnselectedColor();
    }

    public void setIndicatorUnselectedColor(int color) {
        this.pageIndicatorView.setUnselectedColor(color);
    }

    public boolean getScaleOnScroll() {
        return this.scaleOnScroll;
    }

    public void setScaleOnScroll(boolean scaleOnScroll) {
        this.scaleOnScroll = scaleOnScroll;
    }

    public int getSize() {
        return this.size;
    }

    public void setSize(int size) {
        this.size = size;
        this.pageIndicatorView.setCount(size);
    }

    public int getSpacing() {
        return this.spacing;
    }

    public void setSpacing(int spacing) {
        this.spacing = spacing;
    }

    public int getResource() {
        return this.resource;
    }

    public void setResource(int resource) {
        this.resource = resource;
        this.isResourceSet = true;
    }

    public CarouselViewListener getCarouselViewListener() {
        return this.carouselViewListener;
    }

    public void setCarouselViewListener(CarouselViewListener carouselViewListener) {
        this.carouselViewListener = carouselViewListener;
    }

    public CarouselScrollListener getCarouselScrollListener() {
        return this.carouselScrollListener;
    }

    public void setCarouselScrollListener(CarouselScrollListener carouselScrollListener) {
        this.carouselScrollListener = carouselScrollListener;
    }

    private void validate() {
        if (!this.isResourceSet)
            throw new IllegalStateException("Please add a resource layout to populate the CarouselView");
    }

    private IndicatorAnimationType getAnimation(int value) {
        IndicatorAnimationType animationType;
        switch (value) {
            case 1:
                animationType = IndicatorAnimationType.FILL;
                break;
            case 2:
                animationType = IndicatorAnimationType.DROP;
                break;
            case 3:
                animationType = IndicatorAnimationType.SWAP;
                break;
            case 4:
                animationType = IndicatorAnimationType.WORM;
                break;
            case 5:
                animationType = IndicatorAnimationType.COLOR;
                break;
            case 6:
                animationType = IndicatorAnimationType.SCALE;
                break;
            case 7:
                animationType = IndicatorAnimationType.SLIDE;
                break;
            case 8:
                animationType = IndicatorAnimationType.THIN_WORM;
                break;
            case 9:
                animationType = IndicatorAnimationType.SCALE_DOWN;
                break;
            case 0:
            default:
                animationType = IndicatorAnimationType.NONE;
        }
        return animationType;
    }

    private OffsetType getOffset(int value) {
        OffsetType offset;
        switch (value) {
            case 1:
                offset = OffsetType.CENTER;
                break;
            case 0:
            default:
                offset = OffsetType.START;
        }
        return offset;
    }

    public void show() {
        this.validate();
        this.setAdapter();
    }

}

File: carouselview/src/main/java/com/elliecoding/carouselview/CarouselView.kt
package com.elliecoding.carouselview

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SnapHelper
import com.elliecoding.carouselview.control.CarouselLinearLayoutManager
import com.elliecoding.carouselview.control.CarouselSnapHelper
import com.elliecoding.carouselview.control.CarouselViewAdapter
import com.elliecoding.carouselview.enums.IndicatorAnimationType
import com.elliecoding.carouselview.enums.OffsetType
import com.rd.PageIndicatorView
import com.rd.animation.type.AnimationType

// This is a library, external consumers will use the public methods
@Suppress("unused", "MemberVisibilityCanBePrivate")
class CarouselView : FrameLayout {
    private lateinit var indicatorView: PageIndicatorView
    private lateinit var carouselRecyclerView: RecyclerView
    private var layoutManager: CarouselLinearLayoutManager? = null
    private var offsetType: OffsetType? = null
    private var snapHelper: SnapHelper? = null
    private val autoPlayHandler: Handler by lazy {
        Handler(Looper.myLooper()!!)
    }

    var carouselViewListener: CarouselViewListener? = null
    var carouselScrollListener: CarouselScrollListener? = null

    /**
     * Animation type for the indicator below the carousel. Defaults to
     * [IndicatorAnimationType.SLIDE].
     */
    var indicatorAnimationType: IndicatorAnimationType = IndicatorAnimationType.SLIDE
        set(indicatorAnimationType) {
            field = indicatorAnimationType
            when (indicatorAnimationType) {
                IndicatorAnimationType.DROP -> indicatorView.setAnimationType(AnimationType.DROP)
                IndicatorAnimationType.FILL -> indicatorView.setAnimationType(AnimationType.FILL)
                IndicatorAnimationType.NONE -> indicatorView.setAnimationType(AnimationType.NONE)
                IndicatorAnimationType.SWAP -> indicatorView.setAnimationType(AnimationType.SWAP)
                IndicatorAnimationType.WORM -> indicatorView.setAnimationType(AnimationType.WORM)
                IndicatorAnimationType.COLOR -> indicatorView.setAnimationType(AnimationType.COLOR)
                IndicatorAnimationType.SCALE -> indicatorView.setAnimationType(AnimationType.SCALE)
                IndicatorAnimationType.SLIDE -> indicatorView.setAnimationType(AnimationType.SLIDE)
                IndicatorAnimationType.THIN_WORM -> indicatorView.setAnimationType(AnimationType.THIN_WORM)
                IndicatorAnimationType.SCALE_DOWN -> indicatorView.setAnimationType(AnimationType.SCALE_DOWN)
            }
        }

    var enableSnapping = false

    /**
     * Enables or disables autoplay. Changes to this field will be applied immediately, such that
     * you can toggle the feature on or off at any time.
     */
    var autoPlay = false
        set(value) {
            if (autoPlay != value) {
                field = value
                verifyAutoPlay()
            }
        }

    /**
     * Long millisecond value for the time it takes before the autoplay will continue to the next
     * carousel item. It is recommended that you use [java.util.concurrent.TimeUnit.toMillis] to set
     * this value for readability. Setting this to any value <= 0 will set `[autoPlay] = false`
     */
    var autoPlayDelayMillis = 0L
        set(value) {
            field = value
            if (value <= 0L) {
                autoPlay = false
            }
        }
    var scaleOnScroll = false
    var resource = 0
    var size = 0
        set(size) {
            field = size
            indicatorView.count = size
        }
    var spacing = 0
    var currentItem = 0
        set(item) {
            field = if (item < 0) {
                0
            } else if (item >= size) {
                size - 1
            } else {
                item
            }
            carouselRecyclerView.smoothScrollToPosition(currentItem)
        }

    constructor(context: Context) : super(context) {
        init(null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs)
    }

    private fun init(attributeSet: AttributeSet?) {
        val inflater = LayoutInflater.from(context)
        val carouselView = inflater.inflate(R.layout.view_carousel, this)
        carouselRecyclerView = carouselView.findViewById(R.id.carousel_recycler_view)
        indicatorView = carouselView.findViewById(R.id.page_indicator_view)
        carouselRecyclerView.setHasFixedSize(false)
        initializeAttributes(attributeSet)
    }

    private fun initializeAttributes(attributeSet: AttributeSet?) {
        if (attributeSet != null) {
            val attributes = context.theme.obtainStyledAttributes(
                attributeSet,
                R.styleable.CarouselView, 0, 0
            )
            enableSnapping = attributes.getBoolean(R.styleable.CarouselView_enableSnapping, true)
            scaleOnScroll =
                attributes.getBoolean(R.styleable.CarouselView_scaleOnScroll, false)
            autoPlay = attributes.getBoolean(R.styleable.CarouselView_setAutoPlay, false)
            autoPlayDelayMillis =
                attributes.getInteger(R.styleable.CarouselView_setAutoPlayDelay, 2500).toLong()
            carouselOffset =
                getOffset(attributes.getInteger(R.styleable.CarouselView_carouselOffset, 0))
            val resourceId = attributes.getResourceId(R.styleable.CarouselView_resource, 0)
            if (resourceId != 0) {
                resource = resourceId
            }
            val indicatorSelectedColorResourceId =
                attributes.getColor(R.styleable.CarouselView_indicatorSelectedColor, 0)
            val indicatorUnselectedColorResourceId =
                attributes.getColor(R.styleable.CarouselView_indicatorUnselectedColor, 0)
            if (indicatorSelectedColorResourceId != 0) {
                indicatorSelectedColor = indicatorSelectedColorResourceId
            }
            if (indicatorUnselectedColorResourceId != 0) {
                indicatorUnselectedColor = indicatorUnselectedColorResourceId
            }
            indicatorAnimationType = this.getAnimation(
                attributes.getInteger(
                    R.styleable.CarouselView_indicatorAnimationType,
                    0
                )
            )
            indicatorRadius =
                attributes.getInteger(R.styleable.CarouselView_indicatorRadius, 5)
            indicatorPadding =
                attributes.getInteger(R.styleable.CarouselView_indicatorPadding, 5)
            size = attributes.getInteger(R.styleable.CarouselView_size, 0)
            spacing = attributes.getInteger(R.styleable.CarouselView_spacing, 0)
            attributes.recycle()
        }
    }

    fun hideIndicator(hide: Boolean) {
        if (hide) {
            indicatorView.visibility = GONE
        } else {
            indicatorView.visibility = VISIBLE
        }
    }

    override fun onDetachedFromWindow() {
        autoPlay = false
        carouselRecyclerView.clearOnScrollListeners()
        super.onDetachedFromWindow()
    }

    private fun setAdapter() {
        layoutManager = CarouselLinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        layoutManager!!.isOffsetStart(carouselOffset == OffsetType.START)
        if (scaleOnScroll) layoutManager!!.setScaleOnScroll(true)
        carouselRecyclerView.layoutManager = layoutManager
        carouselRecyclerView.adapter = CarouselViewAdapter(
            carouselViewListener,
            resource,
            size,
            carouselRecyclerView,
            spacing,
            carouselOffset == OffsetType.CENTER
        )
        if (enableSnapping) {
            carouselRecyclerView.onFlingListener = null
            snapHelper!!.attachToRecyclerView(carouselRecyclerView)
        }
        setScrollListener()
    }

    private fun setScrollListener() {
        carouselRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                if (layoutManager != null && snapHelper != null) {
                    val centerView = snapHelper!!.findSnapView(layoutManager)
                    if (centerView != null) {
                        setPosition(centerView, recyclerView, newState)
                    }
                }
            }

            private fun setPosition(
                centerView: View,
                recyclerView: RecyclerView,
                newState: Int
            ) {
                val position = layoutManager!!.getPosition(centerView)
                if (carouselScrollListener != null) {
                    carouselScrollListener!!.onScrollStateChanged(
                        recyclerView,
                        newState,
                        position
                    )
                }
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    indicatorView.selection = position
                    currentItem = position
                }
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (carouselScrollListener != null) {
                    carouselScrollListener!!.onScrolled(recyclerView, dx, dy)
                }
            }
        })
    }

    private val autoPlayRunnable = object : Runnable {
        override fun run() {
            if (autoPlay) {
                currentItem = if (size - 1 == currentItem) {
                    0
                } else {
                    currentItem + 1
                }
                autoPlayHandler.postDelayed(this, autoPlayDelayMillis)
            }
        }
    }

    private fun verifyAutoPlay() {
        if (autoPlay) {
            enableAutoPlay()
        } else {
            disableAutoPlay()
        }
    }

    private fun enableAutoPlay() {
        autoPlayHandler.removeCallbacks(autoPlayRunnable)
        autoPlayHandler.postDelayed(autoPlayRunnable, autoPlayDelayMillis)
    }

    private fun disableAutoPlay() {
        autoPlayHandler.removeCallbacks(autoPlayRunnable)
    }

    private var carouselOffset: OffsetType?
        get() = offsetType
        set(offsetType) {
            this.offsetType = offsetType
            snapHelper = when (offsetType) {
                OffsetType.CENTER -> LinearSnapHelper()
                OffsetType.START -> CarouselSnapHelper()
                else -> throw IllegalStateException("Unknown offsetType $offsetType")
            }
        }
    private var indicatorRadius: Int
        get() = indicatorView.radius
        set(radius) {
            indicatorView.radius = radius
        }
    private var indicatorPadding: Int
        get() = indicatorView.padding
        set(padding) {
            indicatorView.padding = padding
        }
    private var indicatorSelectedColor: Int
        get() = indicatorView.selectedColor
        set(color) {
            indicatorView.selectedColor = color
        }
    private var indicatorUnselectedColor: Int
        get() = indicatorView.unselectedColor
        set(color) {
            indicatorView.unselectedColor = color
        }

    private fun validate() {
        check(resource != 0) { "Please add a resource layout to populate the CarouselView" }
    }

    private fun getAnimation(value: Int): IndicatorAnimationType {
        val animationType: IndicatorAnimationType = when (value) {
            1 -> IndicatorAnimationType.FILL
            2 -> IndicatorAnimationType.DROP
            3 -> IndicatorAnimationType.SWAP
            4 -> IndicatorAnimationType.WORM
            5 -> IndicatorAnimationType.COLOR
            6 -> IndicatorAnimationType.SCALE
            7 -> IndicatorAnimationType.SLIDE
            8 -> IndicatorAnimationType.THIN_WORM
            9 -> IndicatorAnimationType.SCALE_DOWN
            0 -> IndicatorAnimationType.NONE
            else -> IndicatorAnimationType.NONE
        }
        return animationType
    }

    private fun getOffset(value: Int): OffsetType {
        val offset: OffsetType = when (value) {
            1 -> OffsetType.CENTER
            0 -> OffsetType.START
            else -> OffsetType.START
        }
        return offset
    }

    fun show() {
        validate()
        setAdapter()
    }
}
